db/db_impl/db_impl.cc

```diff
@@ -3122,38 +2865,29 @@ void DBImpl::PrepareMultiGetKeys(
             CompareKeyContext());
 }
 
-void DBImpl::MultiGet(const ReadOptions& read_options,
-                      ColumnFamilyHandle* column_family, const size_t num_keys,
-                      const Slice* keys, PinnableSlice* values,
-                      Status* statuses, const bool sorted_input) {
-  MultiGet(read_options, column_family, num_keys, keys, values,
-           /* timestamps */ nullptr, statuses, sorted_input);
-}
+void DB::MultiGet(const ReadOptions& _read_options,
+                  ColumnFamilyHandle* column_family, const size_t num_keys,
+                  const Slice* keys, PinnableSlice* values,
+                  std::string* timestamps, Status* statuses,
+                  const bool sorted_input) {
+  // Use an autovector, if possible, to avoid memory allocation overhead
+  autovector<ColumnFamilyHandle*, MultiGetContext::MAX_BATCH_SIZE> cf_autovec;
+  std::vector<ColumnFamilyHandle*> cf_vec;
+  ColumnFamilyHandle** column_families;
 
-void DBImpl::MultiGet(const ReadOptions& _read_options,
-                      ColumnFamilyHandle* column_family, const size_t num_keys,
-                      const Slice* keys, PinnableSlice* values,
-                      std::string* timestamps, Status* statuses,
-                      const bool sorted_input) {
-  if (_read_options.io_activity != Env::IOActivity::kUnknown &&
-      _read_options.io_activity != Env::IOActivity::kMultiGet) {
-    Status s = Status::InvalidArgument(
-        "Can only call MultiGet with `ReadOptions::io_activity` is "
-        "`Env::IOActivity::kUnknown` or `Env::IOActivity::kMultiGet`");
-    for (size_t i = 0; i < num_keys; ++i) {
-      if (statuses[i].ok()) {
-        statuses[i] = s;
-      }
-    }
-    return;
+  if (num_keys > MultiGetContext::MAX_BATCH_SIZE) {
+    cf_vec.resize(num_keys);
+    column_families = cf_vec.data();
+  } else {
+    cf_autovec.resize(num_keys);
+    column_families = &cf_autovec[0];
   }
 
-  ReadOptions read_options(_read_options);
-  if (read_options.io_activity == Env::IOActivity::kUnknown) {
-    read_options.io_activity = Env::IOActivity::kMultiGet;
+  for (size_t i = 0; i < num_keys; ++i) {
+    column_families[i] = column_family;
```

Comment:
You should be able to populate std::vector or std::array with copies of a single value without the need for an explicit loop.